src/Symfony/Component/JsonEncoder/Attribute/DecodeFormatter.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\JsonEncoder\Attribute;
+
+/**
+ * Defines a callable that will be used to format the property data during decoding.
+ *
+ * The first argument of the callable is the input data.
+ * It is possible to inject the configuration using the $config parameter.
+ *
+ * That callable must return the new data.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ *
+ * @experimental
+ */
+#[\Attribute(\Attribute::TARGET_PROPERTY)]
+readonly class DecodeFormatter
+{
+    public \Closure $formatter;
+
+    /**
+     * @param callable(mixed, array=): mixed $formatter
+     */
+    public function __construct(callable $formatter)
```

Comment:
Indeed, you're totally right, I ended up by the following:
- I replaced static callables by service IDs that must refer to `NormalizerInterface` and `DenormalizerInterface` services.
- In the framework integration PR, those normalizers/denormalizers will be put in a service locator thanks to their tag, and they'll be injected in the JsonEncoder/JsonDecoder

Reply:
Maybe can we can we allow more types (services, methods etc) instead of removing support for callable. For simple use cases, callable are convenients.